subsys/net/lib/sockets/socketpair.c

```diff
@@ -0,0 +1,1132 @@
+}
+
+/** Determine if a @ref spair is connected */
+static inline bool sock_is_connected(const struct spair *spair)
+{
+	const struct spair *remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	if (remote == NULL) {
+		return false;
+	}
+
+	return true;
+}
+
+#undef sock_is_eof
+/** Determine if a @ref spair has encountered end-of-file */
+static inline bool sock_is_eof(const struct spair *spair)
+{
+	return !sock_is_connected(spair);
+}
+
+/**
+ * Determine bytes available to write
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * written to a given @ref spair without blocking.
+ */
+static inline size_t spair_write_avail(struct spair *spair)
+{
+	struct spair *const remote = z_get_fd_obj(spair->remote,
+		(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+	struct k_pipe *const p = (remote == NULL)
+		? NULL
+		: &remote->recv_q;
+
+	if (remote == NULL) {
+		return 0;
+	}
+
+	return k_pipe_write_avail(p);
+}
+
+/**
+ * Determine bytes available to read
+ *
+ * Specifically, this function calculates the number of bytes that may be
+ * read from a given @ref spair without blocking.
+ */
+static inline size_t spair_read_avail(struct spair *spair)
+{
+	struct k_pipe *const p = &spair->recv_q;
+
+	return k_pipe_read_avail(p);
+}
+
+/** Swap two 32-bit integers */
+static inline void swap32(u32_t *a, u32_t *b)
+{
+	u32_t c;
+
+	c = *b;
+	*b = *a;
+	*a = c;
+}
+
+/**
+ * Delete @param spair
+ *
+ * This function deletes one endpoint of a socketpair.
+ *
+ * Theory of operation:
+ * - we have a socketpair with two endpoints: A and B
+ * - we have two threads: T1 and T2
+ * - T1 operates on endpoint A
+ * - T2 operates on endpoint B
+ *
+ * There are two possible cases where a blocking operation must be notified
+ * when one endpoint is closed:
+ * -# T1 is blocked reading from A and T2 closes B
+ *    T1 waits on A's write signal. T2 triggers the remote
+ *    @ref spair.write_signal
+ * -# T1 is blocked writing to A and T2 closes B
+ *    T1 is waits on B's read signal. T2 triggers the local
+ *    @ref spair.read_signal.
+ *
+ * If the remote endpoint is already closed, the former operation does not
+ * take place. Otherwise, the @ref spair.remote of the local endpoint is
+ * set to -1.
+ *
+ * If no threads are blocking on A, then the signals have no effect.
+ *
+ * The memeory associated with the local endpoint is cleared and freed.
+ */
+static void spair_delete(struct spair *spair)
+{
+	int res;
+
+	struct spair *remote = NULL;
+	bool have_remote_sem = false;
+
+	if (spair == NULL) {
+		return;
+	}
+
+	if (spair->remote != -1) {
+		remote = z_get_fd_obj(spair->remote,
+			(const struct fd_op_vtable *)&spair_fd_op_vtable, 0);
+
+		if (remote != NULL) {
+			res = k_sem_take(&remote->sem, K_FOREVER);
+			if (res == 0) {
+				have_remote_sem = true;
+				remote->remote = -1;
+				res = k_poll_signal_raise(&remote->write_signal,
+					SPAIR_SIG_CANCEL);
```

Comment:
Indentation off here

Reply:
Chrckpatch was fine with it. What would you prefer?

Reply:
> Chrckpatch was fine with it. What would you prefer?

@cfriedt , we align wrapped function args on the same column as 1st arg (using tabs, then spaces).

Reply:
Keep in mind #21392 and #23823 should eventually make it so that whitespace issues simply don't exist.  It is such a huge waste of reviewer and developer time..

Reply:
I personally don't find it such a big problem (but it took me months to remember conventions of this particular project, though if you just follow pattern of existing code, you'll be fine with any project). I also know that throwing more ~~stupid~~ automated tools will make some aspects worse. But for a project of scale of Zephyr, I agree that's the right direction to tighten up formal codestyle checks, so godspeed with pushing that.